### ISSUE TITLE
Connection leak due to not closing input stream on error

### ISSUE THREAD

**Issue by ben-manes** [googlemaps/google-maps-services-java#179]
> WARN  [2016-09-14 22:31:24,510] com.squareup.okhttp.OkHttpClient: A connection to https://maps.googleapis.com/ was leaked. Did you forget to close a response body?

`OkHttpPendingResult#getBytes()` does not close input stream. While OkHttp can recover on a 2xx, it will not do this for you on an error code. As [explained](https://github.com/square/okhttp/issues/2311), they expect callers to write proper resource management code.

When a failure occurs this library typically hangs, as seen in #161, and requires callers perform their own timeouts. Unfortunately calling `PendingResult#cancel()` afterwards is not good enough to recover, because the connection is stuck by the open input stream. Eventually the pool will be exhausted and no new requests are performed. The only resolution is to restart the application.

**Comment by shmuelr** [comment]
Do you think updating to OkHttp 3.x help this issue?

**Comment by ben-manes** [comment]
I think closing the input stream, as is standard practice, would be best.

**Comment by domesticmouse** [comment]
@ben-manes I'm currently trying to do a few too many things at once, so if you can dive in and propose a fix via a pull request, that would be super helpful. Thanks!

**Comment by IvanRF** [comment]
@ben-manes I'm getting the same warning but over `com.squareup.okhttp.ConnectionPool`:

```
mar 07, 2017 8:17:10 PM com.squareup.okhttp.ConnectionPool pruneAndGetAllocationCount
WARNING: A connection to https://maps.googleapis.com/ was leaked. Did you forget to close a response body?
```

I get that every now and then when using `GeocodingApi.geocode(GeoApiContext, String)`.

**Comment by ben-manes** [comment]
Yes, that was what I [found](https://github.com/googlemaps/google-maps-services-java/issues/161#issuecomment-250562574), too.

I had to drop this library as unstable. Instead I use its serializers with Retrofit to make the api calls. That was fast to implement and I haven't had any problems since.

**Comment by IvanRF** [comment]
Oh, ok, thanks for the fast reply 👍

**Comment by esetnik** [comment]
@ben-manes @IvanRF i'm having the exact same issue. This needs to be reopened.

**Comment by unitydynamics** [comment]
Having the same issue here.  Please put this at the top of the list!  Library is not stable.

**Comment by esetnik** [comment]
@ben-manes care to share your alternate implementation? I'm also using retrofit and i'd love to piggyback on the work you've already done.

**Comment by ben-manes** [comment]
See this [comment](https://github.com/googlemaps/google-maps-services-java/issues/161#issuecomment-251040776) and register the serializers from the given class.

**Comment by domesticmouse** [comment]
@esetnik Can you please share a repro case that I can test with? I'm in the process of migrating to OkHttp3, and closing up issues where I can.

**Comment by esetnik** [comment]
@domesticmouse I don't have a public repo available that reproduces the issue and tbh i'm not exactly sure how to reproduce it. I only notice it happening when sentry reports errors in our infrastructure. It is relatively rare (occurring only a few times per day with ~100k google maps api calls). I switched to the retrofit implementation provided by @ben-manes and haven't had any issues since.

**Comment by domesticmouse** [comment]
Thanks @esetnik. Given that, I'm going to close this bug for now.

**Comment by esetnik** [comment]
@domesticmouse I believe I have fixed this issue in #409.
